[GEOT-5623]: dynamic stroke-dasharray improvements

wumpz opened the pull request:
Hopefully, now I got the EOLs correct. 

This patch makes GeoTools able to process a complete stroke-dasharray from database. 

The implementation of GeoTools so far relies on getting single numbers from a stroke-dasharray from expression. Internally a stroke-dasharray is converted to a list of expressions. For this type of stroke-dasharray there must be more than one expression to be correct.

This patch introduces for the "edge" case "only one expression for a stroke-dasharray" to get the whole dasharray from one expression.

wumpz commented:
Is it now reviewable?

aaime commented:
It is against the wrong branch, this knid of change will be accepted only on master, and has a bit of reformats that should not be there, but yes, it's reviewable, I'd wait for a review about its actual contents before removing the reformats and targeting the master branch.

nmco commented:
Starting to review this pull request content. Don't forget to update the issue [GEOT-3776](https://osgeo-org.atlassian.net/browse/GEOT-5623), assign it to yourself, put status in progress and once this is merged update the fixed versions.

nmco commented on modules/library/main/src/main/java/org/geotools/styling/visitor/RescaleStyleVisitor.java:
```diff
@@ -64,91 +64,117 @@
+            }
+        }
+        return rescaled;
+    }
```
You could apply the rescaling pattern to both uses cases, something like this:

```java
protected List<Expression> rescaleDashArray(List<Expression> expressions) {
    if (expressions == null || expressions.isEmpty()) {
        return expressions;
    }
    List<Expression> rescaled = new ArrayList<>(expressions.size());
    for (Expression expression : expressions) {
        Expression rescale = ff.function("listMultiply", scale, expression);
        if (expression instanceof Literal && scale instanceof Literal) {
            String constant = (String) rescale.evaluate(null);
            rescaled.add(ff.literal(constant));
        }
    }
    return rescaled;
}
```

nmco commented on modules/library/main/src/main/java/org/geotools/styling/visitor/UomRescaleStyleVisitor.java:
```diff
@@ -98,20 +98,28 @@ protected Expression rescale(Expression unscaled, Unit<Length> uom) {
+            }
         }
         return rescaled;
     }
```
Same as above you could apply the same rescaling process for both.

nmco commented:
I'm not very comfortable with the ``listMultiply`` function, the name of the function makes it looks like a generic function capable of multiplying a list of numbers by a constant.

But in really this is a function that takes a string representing a list of numbers or a number convertible to string and multiplies them by a factor and convert them back to a string.

I cannot propose a better alternative so if no one else complains I would say let it go, in the future someone may rename it or extend it as needed.

nmco commented:
If have finish this first review. Tests for all the new functionalities were added, well done :+1:. No explicit documentation is required for this changes.

wumpz commented on modules/library/render/src/main/java/org/geotools/renderer/style/SLDStyleFactory.java:
```diff
@@ -1036,6 +1028,30 @@ Stroke getStroke(org.geotools.styling.Stroke stroke, Object feature) {
+            }
+        }
+        return dashes;
+    }
```
Sure one could generalize this method. My problem is performance. The original way should be faster. Thats why I decided to make it a special case. But if you prefer to process all possibilities like so I will change it.

nmco replied:
By generalization I mean that this use case will also be supported (is not just a code style issue):
```
<CssParameter name="stroke-dasharray">
	<ogc:Literal>1.0 4.5</ogc:Literal>
	<ogc:Literal>2.0 1.0</ogc:Literal>
</CssParameter>
```

wumpz commented:
So how should I integrate the changes? Simple additional commits to my pull request?

nmco commented:
As Andrea say above, this pull request is against the wrong branch. Pull request should always be made first against master and only after backported.

Note that new features have a one month cool down rule before being allowed to be backported, in any case you should ask permission on the mailing list before backporting a new feature.

I invite you to read the GeoTools [project procedures](http://docs.geotools.org/latest/developer/procedures/index.html).

In your case you will also need to get ride of the extra reformats. 

In your case I will proceed as follow (git allow you to reach this goals in different ways this just an example). I will assume that you are working on a fork of GeoTools and added the original GeoTools as ``upstream``:

- [create a new branch from GeoTools master] 
  ``get checkout -b "dynamic-dash-array" upstream/master``

- [cherry pick this work]
  ``git cherry-pick 55d8e360806a3291430f3f2c60fb3fbf554bf836``

- [prepare to clean the extra reformats, this will allow you to see your changes]
  ``got reset --soft HEAD~1``
  ``git reset HEAD *``

- [once you get ride of extra reformats and fixed everything just do a pull request]
 ``git add .``
 ``git commit -am "[GEOT-5623]: dynamic stroke-dasharray improvements"``
 ``git push -u origin dynamic-dash-array``

You situation is "special" because you need to get ride of the extra reformats. A typical work-flow for fixing a pull request and contribute back is:

- [do your changes and commit]
  ``git commit -am "my new changes"

- [now you should squash your commits, only do this in branches where you are the only one working]
  ``git rebase --interactive HEAD~2``

- [since you rewrite you commit history you will need to force your push]
  ``git push -f``

Git ``squash`` operation is very useful for fixing pull requests but is also a very powerful operation use it with caution.

wumpz commented:
That was clear. But listMultiply comes surely with a performance impact. So should it be used in general?

-------- Ursprüngliche Nachricht --------
Von: Nuno Oliveira <notifications@github.com> 
Datum: 13.02.2017  22:48  (GMT+01:00) 
An: geotools/geotools <geotools@noreply.github.com> 
Cc: Tobias <t.warneke@gmx.net>, Author <author@noreply.github.com> 
Betreff: Re: [geotools/geotools] [GEOT-5623]: dynamic stroke-dasharray
  improvements (#1467) 

@nmco commented on this pull request.



In modules/library/render/src/main/java/org/geotools/renderer/style/SLDStyleFactory.java:
> +                    dashes = new float[strok.length];
+                    for (int l = 0; l < dashes.length; l++) {
+                        dashes[l] = Float.parseFloat(strok[l]);
+                    }
+                }
+            } else {
+                dashes = new float[stroke.dashArray().size()];
+                int index = 0;
+                for (Expression expression : stroke.dashArray()) {
+                    dashes[index] = expression.evaluate(feature, Float.class);
+                    index++;
+                }
+            }
+        }
+        return dashes;
+    }

By generalization I mean that this use case will also be supported (is not just a code style issue):
<CssParameter name="stroke-dasharray">
	<ogc:Literal>1.0 4.5</ogc:Literal>
	<ogc:Literal>2.0 1.0</ogc:Literal>
</CssParameter>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/geotools/geotools","title":"geotools/geotools","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/geotools/geotools"}},"updates":{"snippets":[{"icon":"PERSON","message":"@nmco commented on #1467"}],"action":{"name":"View Pull Request","url":"https://github.com/geotools/geotools/pull/1467"}}}​​​​​

wumpz commented:
Why should the reformats removed, as it's clearly improving format by removing unnecessary spaces at lineend. Nevertheless it was unintended. 

-------- Ursprüngliche Nachricht --------
Von: Nuno Oliveira <notifications@github.com> 
Datum: 13.02.2017  22:38  (GMT+01:00) 
An: geotools/geotools <geotools@noreply.github.com> 
Cc: Tobias <t.warneke@gmx.net>, Author <author@noreply.github.com> 
Betreff: Re: [geotools/geotools] [GEOT-5623]: dynamic stroke-dasharray
  improvements (#1467) 

As Andrea say above, this pull request is against the wrong branch. Pull request should always be made first against master and only after backported.
Note that new features have a one month cool down rule before being allowed to be backported, in any case you should ask permission on the mailing list before backporting a new feature.
I invite you to read the GeoTools project procedures.
In your case you will also need to get ride of the extra reformats.
In your case I will proceed as follow (git allow you to reach this goals in different ways this just an example). I will assume that you are working on a fork of GeoTools and added the original GeoTools as upstream:


[create a new branch from GeoTools master]

get checkout -b "dynamic-dash-array" upstream/master


[cherry pick this work]

git cherry-pick 55d8e360806a3291430f3f2c60fb3fbf554bf836


[prepare to clean the extra reformats, this will allow you to see your changes]

got reset --soft HEAD~1

git reset HEAD *


[once you get ride of extra reformats and fixed everything just do a pull request]

git add .

git commit -am "[GEOT-5623]: dynamic stroke-dasharray improvements"

git push -u origin dynamic-dash-array


You situation is "special" because you need to get ride of the extra reformats. A typical work-flow for fixing a pull request and contribute back is:


[do your changes and commit]

``git commit -am "my new changes"


[now you should squash your commits, only do this in branches where you are the only one working]

git rebase --interactive HEAD~2


[since you rewrite you commit history you will need to force your push]

git push -f


Git squash operation is very useful for fixing pull requests but is also a very powerful operation use it with caution.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/geotools/geotools","title":"geotools/geotools","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/geotools/geotools"}},"updates":{"snippets":[{"icon":"PERSON","message":"@nmco in #1467: As Andrea say above, this pull request is against the wrong branch. Pull request should always be made first against master and only after backported.\r\n\r\nNote that new features have a one month cool down rule before being allowed to be backported, in any case you should ask permission on the mailing list before backporting a new feature.\r\n\r\nI invite you to read the GeoTools [project procedures](http://docs.geotools.org/latest/developer/procedures/index.html).\r\n\r\nIn your case you will also need to get ride of the extra reformats. \r\n\r\nIn your case I will proceed as follow (git allow you to reach this goals in different ways this just an example). I will assume that you are working on a fork of GeoTools and added the original GeoTools as ``upstream``:\r\n\r\n- [create a new branch from GeoTools master] \r\n  ``get checkout -b \"dynamic-dash-array\" upstream/master``\r\n\r\n- [cherry pick this work]\r\n  ``git cherry-pick 55d8e360806a3291430f3f2c60fb3fbf554bf836``\r\n\r\n- [prepare to clean the extra reformats, this will allow you to see your changes]\r\n  ``got reset --soft HEAD~1``\r\n  ``git reset HEAD *``\r\n\r\n- [once you get ride of extra reformats and fixed everything just do a pull request]\r\n ``git add .``\r\n ``git commit -am \"[GEOT-5623]: dynamic stroke-dasharray improvements\"``\r\n ``git push -u origin dynamic-dash-array``\r\n\r\nYou situation is \"special\" because you need to get ride of the extra reformats. A typical work-flow for fixing a pull request and contribute back is:\r\n\r\n- [do your changes and commit]\r\n  ``git commit -am \"my new changes\"\r\n\r\n- [now you should squash your commits, only do this in branches where you are the only one working]\r\n  ``git rebase --interactive HEAD~2``\r\n\r\n- [since you rewrite you commit history you will need to force your push]\r\n  ``git push -f``\r\n\r\nGit ``squash`` operation is very useful for fixing pull requests but is also a very powerful operation use it with caution.\r\n\r\n\r\n"}],"action":{"name":"View Pull Request","url":"https://github.com/geotools/geotools/pull/1467#issuecomment-279530703"}}}

nmco commented:
The two main reasons for not accepting extra reformats in pull requests are:
* making it harder than necessary to review pull request
* making it harder to check history about a certain line of code

There is a n ongoing discussion about this on the mailing list:
http://osgeo-org.1560.x6.nabble.com/Formatting-formatting-getting-tired-of-all-complaints-td5307518.html

nmco commented:
Regarding the performance penalty of the generalized method:
* arguing about performance problems without a proper benchmark is not going anywhere
* in this situation we are arguing about the cost of an extra split versus having a "big hack" in the code

That said, I made a quick micro benchmark to have an idea of the cost of the extra split, 10 millions runs where made for each case (with proper warm up, repetitions, etc ...):
* without split - 10 millions runs - total time 1456 milliseconds
* with extra split - 10 millions runs - total time 3717 milliseconds

My interpretation of this is that when rendering 10000 features the cost of the extra split will be around 1 millisecond.

So as far has I'm concerned I think hat the generalized method should be used.

So yes I think that the generalized method should be used. Feel free to push this discussion the mailing list if you think is necessary.

wumpz commented:
Opened a new pull request: #1492
